pkg/server/commands/listusers/list_users_rpc.go

```diff
@@ -216,26 +217,36 @@ func (l *listUsersQuery) ListUsers(
 	go func() {
 		internalRequest := fromListUsersRequest(req, &datastoreQueryCount, &dispatchCount)
 		resp := l.expand(cancellableCtx, internalRequest, foundUsersCh)
-		// first send error and then close results channel, to ensure that error takes precedence
 		if resp.err != nil {
 			expandErrCh <- resp.err
 		}
 		close(foundUsersCh)
 	}()
 
+	deadlineExceeded := false
+
 	select {
-	// Note: if all cases can proceed, one will be selected at random
-	case err := <-expandErrCh:
-		telemetry.TraceError(span, err)
-		return nil, err
 	case <-doneWithFoundUsersCh:
 		break
 	case <-cancellableCtx.Done():
+		deadlineExceeded = true
 		// to avoid a race on the 'foundUsersUnique' map below, wait for the range over the channel to close
 		<-doneWithFoundUsersCh
 		break
 	}
 
+	select {
+	case err := <-expandErrCh:
+		if deadlineExceeded || errors.Is(err, context.DeadlineExceeded) {
+			// We skip the error because we want to send at least partial results to the user (but we should probably set response headers)
+			break
+		}
+		telemetry.TraceError(span, err)
+		return nil, err
+	default:
+		break
+	}
+
```

Comment:
This second `select`  statement is blocked either by a value on the `doneWithFoundUsersCh` or context cancellation. If an error gets emitted, I'm not sure that either that either will get triggered. This means that even if an error gets emitted, we're blocking until the context cancels. Or so that's my understanding currently. 

Can you demonstrate with a test that we fail as soon as the error gets emitted; that we're not always waiting for context to cancel?

Reply:
> If an error gets emitted, I'm not sure that either that either will get triggered.

The line `case <-doneWithFoundUsersCh:` will be executed after `foundUsersCh` is closed, which is after `expandErrCh <- resp.err`

Reply:
>The line case <-doneWithFoundUsersCh: will be executed after foundUsersCh is closed, which is after expandErrCh <- resp.err.

Ok I see this now, makes sense.

I'd also like to double-check (ideally with a test) that the appropriate tests pass when setting a deadline to `0`.

Reply:
Added `infinite_deadline_does_not_block_return_of_errors`